ember-core/shared/src/main/scala/org/http4s/ember/core/h2/H2Connection.scala

```diff
@@ -192,12 +192,15 @@ private[h2] class H2Connection[F[_]](
     firstGoAway.getOrElse(F.unit) >> go(chunk)
   }
 
-  def writeLoop: Stream[F, Nothing] =
-    Stream
-      .fromQueueUnterminatedChunk[F, H2Frame](outgoing, Int.MaxValue)
-      .chunks
-      .foreach(writeChunk)
-      .handleErrorWith(ex => Stream.exec(logger.debug(ex)("writeLoop terminated")))
+  def writeLoop: F[Unit] = {
+    // the first `take` will semantically block until items are available
+    // NOTE: perhaps we should add a limit to number of frames, albeit high,
+    // if only to avoid having a space leak in here.
+    val cycle = outgoing.take.flatMap(writeChunk) >>
+      outgoing.tryTakeN(None).flatMap(_.traverse_(writeChunk))
+    cycle.foreverM.handleErrorWith(ex => logger.debug(ex)("writeLoop terminated"))
+  }
```

Comment:
Yes, we definitely need a limit. As-written we have completely lost the backpressure signal.

Reply:
Ok. Note that the lack of limit (or rather the use of Int.Max_Value) was already there in the existing code. This PR is not making this aspect worse.

Reply:
What would a reasonable number of frames be? 1024? Note that this is just how many frames it takes off in each trip to the queue, so as long as it is large enough to make those infrequent enough... Is there any idea as to how many frames would an Http2  connection emit?

Reply:
Out of curiosity, what happens if we choose too low of a number, like `2`?

We don't lose correctness, we just tank performance by semantically blocking on the queue considerably more?

Reply:
> We don't lose correctness, we just tank performance by semantically blocking on the queue considerably more?

The limit in question is not the size of the queue. It is the parameter to [`tryTakeN`](https://github.com/typelevel/cats-effect/blob/e1ab2aacbd775f0a9981c5937d1bd8468d4ebc42/std/shared/src/main/scala/cats/effect/std/PQueue.scala#L263), which says how many items each cycle of the loop takes from the queue. 

It is not a matter of "semantically blocking". If the write loop (and the downstream write) can write faster that producers put into the queue, it will have to block regardless, which is good. It is about reducing contention on the queue and its synchronisation mechanisms: the more elements we allow picking in each visit, the fewer visits are needed.

The problem, as @armanbilge points out, is that the queue serves not just as a store of pending to process items but also as a "backpressure signal", to indicate to the producers upstream, and to the system as a whole, that the system is getting saturated. If we were to allow too high a limit (or no limit as the current code does) then it could happen that a run of the cycle would sweep a full queue into the local memory of this write loop, hiding the fact that there are a lot of pending work. 

Perhaps allowig up to 16 items should be efficient enough but not put too many items out of view.